### PR TITLE
[SPARK-33557][CORE][MESOS][TEST] Ensure the relationship between STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT and NETWORK_TIMEOUT

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
As described in SPARK-33557, `HeartbeatReceiver` and `MesosCoarseGrainedSchedulerBackend` will always use `Network.NETWORK_TIMEOUT.defaultValueString` as value of `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` when we configure `NETWORK_TIMEOUT` without configure `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT`, this is different from the relationship described in `configuration.md`.

To fix this problem，the main change of this pr as follow:

- Remove the explicitly default value of `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT`

- Use actual value of `NETWORK_TIMEOUT` as `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` when `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` not configured in `HeartbeatReceiver` and `MesosCoarseGrainedSchedulerBackend` 
 

### Why are the changes needed?
To ensure the relationship between `NETWORK_TIMEOUT` and  `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` as we described in `configuration.md`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass the Jenkins or GitHub Action

- Manual test configure `NETWORK_TIMEOUT` and `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` locally

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @HyukjinKwon this pr this bug fix  for SPARK-33557.

By the way, do you know where to use the configuration item `spark.core.connection.ack.wait.timeout` in Spark code? 

https://github.com/apache/spark/blob/6e5446e61f278e9afac342e8f33905f5630aa7d5/core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala#L98-L99

It seems only configured in the above test case

File: repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala [hunk 1]
```diff
@@ -231,6 +231,7 @@ class ExecutorClassLoaderSuite
       .setMaster("local")
       .setAppName("executor-class-loader-test")
       .set("spark.network.timeout", "11s")
+      .set("spark.network.timeoutInterval", "11s")
```
Comment (LuciferYang) [inline repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala:234]: `spark.network.timeoutInterval` default value is 60s, `spark.network.timeoutInterval` should <= `spark.network.timeout`

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala [hunk 2]
```diff
@@ -80,7 +80,7 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
   // executor ID -> timestamp of when the last heartbeat from this executor was received
   private val executorLastSeen = new HashMap[String, Long]
 
-  private val executorTimeoutMs = sc.conf.get(config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT)
+  private val executorTimeoutMs = Utils.blockManagerHeartbeatTimeoutAsMs(sc.conf)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:83]: Can you use `fallbackConf` at the definition of this configuration at `config/package.scala`?

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:83]: definition `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` from 
```
  private[spark] val STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT =
    ConfigBuilder("spark.storage.blockManagerHeartbeatTimeoutMs")
      .version("0.7.0")
      .withAlternative("spark.storage.blockManagerSlaveTimeoutMs")
      .timeConf(TimeUnit.MILLISECONDS)
      .createWithDefaultString(Network.NETWORK_TIMEOUT.defaultValueString)
```
to 
```
private[spark] val STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT =
    ConfigBuilder("spark.storage.blockManagerHeartbeatTimeoutMs")
      .version("0.7.0")
      .withAlternative("spark.storage.blockManagerSlaveTimeoutMs")
      .fallbackConf(Network.NETWORK_TIMEOUT)
```
?

If so, the TimeUnit of `STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT` will change from `TimeUnit.MILLISECONDS` to `TimeUnit.SECONDS`, and the default value will change from `120000` to `120`,
This does not match the "Ms" unit described in `spark.storage.blockManagerHeartbeatTimeoutMs`.

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:83]: So if use `fallbackConf `, we need to change the name of `spark.storage.blockManagerHeartbeatTimeoutMs` to remove the suffix "Ms".

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:83]: Could we maybe do something like this?

```diff
diff --git a/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala b/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala
index 233ad884a72..9aabe46e721 100644
--- a/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala
+++ b/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala
@@ -80,7 +80,9 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
   // executor ID -> timestamp of when the last heartbeat from this executor was received
   private val executorLastSeen = new HashMap[String, Long]

-  private val executorTimeoutMs = sc.conf.get(config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT)
+  private val executorTimeoutMs = sc.conf.get(
+    config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
+  ).getOrElse(sc.conf.getTimeAsMs(Network.NETWORK_TIMEOUT.key))

   private val checkTimeoutIntervalMs = sc.conf.get(Network.NETWORK_TIMEOUT_INTERVAL)

diff --git a/core/src/main/scala/org/apache/spark/internal/config/package.scala b/core/src/main/scala/org/apache/spark/internal/config/package.scala
index 093a0ecf58d..f42388e24ce 100644
--- a/core/src/main/scala/org/apache/spark/internal/config/package.scala
+++ b/core/src/main/scala/org/apache/spark/internal/config/package.scala
@@ -504,7 +504,7 @@ package object config {
       .version("0.7.0")
       .withAlternative("spark.storage.blockManagerSlaveTimeoutMs")
       .timeConf(TimeUnit.MILLISECONDS)
-      .createWithDefaultString(Network.NETWORK_TIMEOUT.defaultValueString)
+      .createOptional

   private[spark] val STORAGE_CLEANUP_FILES_AFTER_EXECUTOR_EXIT =
     ConfigBuilder("spark.storage.cleanupFilesAfterExecutorExit")
diff --git a/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala b/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala
index efcef09132f..88fc8b29d35 100644
--- a/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala
+++ b/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala
@@ -651,7 +651,9 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
           .registerDriverWithShuffleService(
             agent.hostname,
             externalShufflePort,
-            sc.conf.get(config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT),
+            sc.conf.get(
+              config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
+            ).getOrElse(sc.conf.getTimeAsMs(Network.NETWORK_TIMEOUT.key)),
             sc.conf.get(config.EXECUTOR_HEARTBEAT_INTERVAL))
         agent.shuffleRegistered = true
       }
```

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -2977,6 +2977,14 @@ private[spark] object Utils extends Logging {
       0
     }
   }
+
+  /**
+   * Returns blockmanager heartbeat timeout as milliseconds.
+   */
+  def blockManagerHeartbeatTimeoutAsMs(conf: SparkConf): Long = {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:2984]: I think you don't have to bother with this util and test. You can just write down, in the PR description,  that you tested it like this locally

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2984]: OK

**Comment by HyukjinKwon** [comment]
Oh, let's don't forgot to update the PR description as well.

File: repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala:234]: without this line , this suite failed cause by :
```
requirement failed: spark.network.timeoutInterval should be less than or equal to spark.storage.blockManagerHeartbeatTimeoutMs.
java.lang.IllegalArgumentException: requirement failed: spark.network.timeoutInterval should be less than or equal to spark.storage.blockManagerHeartbeatTimeoutMs.

```

File: core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala [hunk 4]
```diff
@@ -80,7 +80,9 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
   // executor ID -> timestamp of when the last heartbeat from this executor was received
   private val executorLastSeen = new HashMap[String, Long]
 
-  private val executorTimeoutMs = sc.conf.get(config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT)
+  private val executorTimeoutMs = sc.conf.get(
+    config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
+  ).getOrElse(Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_TIMEOUT)}s"))
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:85]: There are two reasons for change from 
```
sc.conf.get(
  config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
).getOrElse(sc.conf.getTimeAsMs(Network.NETWORK_TIMEOUT.key))
```
to 

```
sc.conf.get(
    config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
  ).getOrElse(Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_TIMEOUT)}s"))
```
：

- Maybe `Network.NETWORK_TIMEOUT` is not configured, need a default value

- TimeUnit of `Network.NETWORK_TIMEOUT` is `TimeUnit.SECONDS` , so if user configure  `Network.NETWORK_TIMEOUT`  to `200` not `200s`, the result of `executorTimeoutMs` will be `200ms` not `200000ms` when use `sc.conf.getTimeAsMs(Network.NETWORK_TIMEOUT.key)`, it should be wrong

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:85]: Yeah, that's fine.

**Review by HyukjinKwon — APPROVED** [review]
LGTM2

**Comment by HyukjinKwon** [comment]
Hm, the test failures look consistent?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I retriggered GitHub Actions too.

**Comment by LuciferYang** [comment]
4 test cases of the yarn module failed in the GitHub action：

```
YarnClusterSuite.run Spark in yarn-client mode with different configurations, ensuring redaction
YarnClusterSuite.run Spark in yarn-cluster mode with different configurations, ensuring redaction
YarnClusterSuite.yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630)
YarnClusterSuite.run Spark in yarn-client mode with additional jar
```

but local test success:

```
Run completed in 8 minutes, 56 seconds.
Total number of tests run: 137
Suites: completed 18, aborted 0
Tests: succeeded 137, failed 0, canceled 1, ignored 0, pending 0
All tests passed.
```
Let me check kafka-sql test failed in Jenkins.

**Comment by LuciferYang** [comment]
All failed cases in Jenkins belong to kafka10-sql module, but local test succcess:
```
mvn clean install -pl external/kafka-0-10-sql 

Run completed in 11 minutes, 42 seconds.
Total number of tests run: 260
Suites: completed 26, aborted 0
Tests: succeeded 260, failed 0, canceled 0, ignored 4, pending 0
```

**Comment by LuciferYang** [comment]
GitHub Action passed except `Pull Request Labeler / Report test results (pull_request_target) `, looks like it's the result of a previous test

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, this is a valid PR. I made a backporting PR.
- https://github.com/apache/spark/pull/31175

**Comment by LuciferYang** [comment]
thx~ @dongjoon-hyun

**Pull request closed without merging**